docs/proposals/volume-provisioning.md

```diff
@@ -125,30 +130,270 @@ We propose that:
-    `ProvisionableVolumePlugin`, which has a method called `NewProvisioner`
-    that returns a new provisioner.
+  * The provisioner SHOULD send events on a claim to report any errors
+    related to provisioning a volume for the claim. This way, users get the same
+    experience as with internal provisioners.
 
-6.  The controller calls volume plugin `Provision` with Parameters from the `StorageClass` configuration object.
+  * The provisioner MUST implement also a deleter. It must be able to delete
+    storage assets it created. It MUST NOT assume that any other internal or
+    external plugin is present.
 
-7.  If `Provision` returns an error, the controller generates an event on the
-    claim and goes back to step 1., i.e. it will retry provisioning periodically
+  The external provisioner runs in a separate process which watches claims, be
+  it an external storage appliance, a daemon or a Kubernetes pod. For every
+  claim creation or update, it implements these steps:
 
-8.  If `Provision` returns no error, the controller creates the returned
-    `api.PersistentVolume`, fills its `Class` attribute with `claim.Spec.Class`
-    and makes it already bound to the claim
+  1. The provisioner inspects if
+     `claim.Annotations["volume.beta.kubernetes.io/storage-provisioner"] == <provisioner name>`.
+     All other claims MUST be ignored.
 
-  1.  If the create operation for the `api.PersistentVolume` fails, it is
-      retried
+  2. The provisioner MUST check that the claim is unbound, i.e. its
+     `claim.Spec.VolumeName` is empty. Bound volumes MUST be ignored.
 
-  2.  If the create operation does not succeed in reasonable time, the
-      controller attempts to delete the provisioned volume and creates an event
-      on the claim
+     *Race condition when the provisioner provisions a new PV for a claim and
+     at the same time Kubernetes binds the same claim to another PV that was
+     just created by admin is discussed below.*
 
-Existing behavior is un-changed for claims that do not specify `claim.Spec.Class`.
+  3. It tries to find a StorageClass instance referenced by annotation
+     `claim.Annotations["volume.beta.kubernetes.io/storage-class"]`. If not
+     found, it SHOULD report an error (by sending an event to the claim) and it
+     SHOULD retry periodically with step i.
+
+  4. The provisioner MUST parse arguments in the `StorageClass` and
+     `claim.Spec.Selector` and provisions appropriate storage asset that matches
+     both the parameters and the selector.
+     When it encounters unknown parameters in `storageClass.Parameters` or
+     `claim.Spec.Selector` or the combination of these parameters is impossible
+     to achieve, it SHOULD report an error and it MUST NOT provision a volume.
+     All errors found during parsing or provisioning SHOULD be send as events
+     on the claim and the provisioner SHOULD retry periodically with step i.
+
```

Comment:
You are right, the text is different for internal and external provisioners, however the result is the same: if there is a selector in a PVC, the provisioner must provision something that matches both the class and the selector. In-tree provisioners are stupid and don't parse selectors right now (this will change in future, 1.6?) and therefore refuse to provision anything with any selector. I would advise external provisoners do the same - either refuse to provision a claim with *any* selector or explicitly select few labels that are allowed in a PVC selector and document those as additional way how to configure a volume. Therefore "MUST parse claim.Spec.Selector" may be just `if pvc.Spec.Selector != nil { return Error("can't parse selector!") }`

I'll try to capture this idea in the proposal...